11_FFTBloom/main.cpp

```diff
@@ -0,0 +1,1192 @@
+		//       and setting stuff via shader push constants (such as which axis to perform FFT on and the size of output image).
+
+		// -------------------------------------- KERNEL FFT PRECOMP ----------------------------------------------------------------
+		{
+			// Pipeline Layouts
+			smart_refctd_ptr<IGPUPipelineLayout> kernelFirstAxisFFTPipelineLayout;
+			{
+				auto sampler = createSampler(ISampler::E_TEXTURE_CLAMP::ETC_CLAMP_TO_BORDER);
+				IGPUDescriptorSetLayout::SBinding bnd =
+				{
+					IDescriptorSetLayoutBase::SBindingBase(),
+					0u,
+					IDescriptor::E_TYPE::ET_COMBINED_IMAGE_SAMPLER,
+					IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+					IShader::E_SHADER_STAGE::ESS_COMPUTE,
+					1u,
+					&sampler
+				};
+
+				kernelFirstAxisFFTPipelineLayout = createPipelineLayout({ &bnd, 1 });
+			}
+
+			smart_refctd_ptr<IGPUPipelineLayout> kernelNormalizationPipelineLayout;
+			{
+				IGPUDescriptorSetLayout::SBinding bnd =
+				{
+					IDescriptorSetLayoutBase::SBindingBase(),
+					0u,
+					IDescriptor::E_TYPE::ET_STORAGE_IMAGE,
+					IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+					IShader::E_SHADER_STAGE::ESS_COMPUTE,
+					CHANNELS,
+					nullptr
+				};
+
+				kernelNormalizationPipelineLayout = createPipelineLayout({ &bnd, 1 });
+			}
+
+
+			const asset::VkExtent3D paddedKerDim = padDimensions(kerDim);
+
+			// create kernel spectrums
+			auto createKernelSpectrum = [&]() -> auto
+				{
+					video::IGPUImage::SCreationParams imageParams;
+					imageParams.flags = static_cast<asset::IImage::E_CREATE_FLAGS>(0u);
+					imageParams.type = asset::IImage::ET_2D;
+					imageParams.format = useHalfFloats ? EF_R16G16_SFLOAT : EF_R32G32_SFLOAT;
+					imageParams.extent = { paddedKerDim.width,paddedKerDim.height,1u };
+					imageParams.mipLevels = 1u;
+					imageParams.arrayLayers = 1u;
+					imageParams.samples = asset::IImage::ESCF_1_BIT;
+					imageParams.usage = IImage::EUF_STORAGE_BIT | IImage::EUF_SAMPLED_BIT;
+
+					auto kernelImg = m_device->createImage(std::move(imageParams));
+
+					auto memReqs = kernelImg->getMemoryReqs();
+					memReqs.memoryTypeBits &= m_device->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+					auto gpuMem = m_device->allocate(memReqs, kernelImg.get());
+
+					video::IGPUImageView::SCreationParams viewParams;
+					viewParams.flags = static_cast<video::IGPUImageView::E_CREATE_FLAGS>(0u);
+					viewParams.image = kernelImg;
+					viewParams.viewType = video::IGPUImageView::ET_2D;
+					viewParams.format = useHalfFloats ? EF_R16G16_SFLOAT : EF_R32G32_SFLOAT;
+					viewParams.components = {};
+					viewParams.subresourceRange.aspectMask = IImage::EAF_COLOR_BIT;
+					viewParams.subresourceRange.baseMipLevel = 0;
+					viewParams.subresourceRange.levelCount = 1;
+					viewParams.subresourceRange.baseArrayLayer = 0;
+					viewParams.subresourceRange.layerCount = 1;
+					return m_device->createImageView(std::move(viewParams));
+				};
```

Comment:
1. `padDimensions` far away in a header, the only reason for it to be that way if its in a general `hlsl::fft` header
2. why is the spectrum image a full square, its a real signal, it should be `padDimensions.someComponent/2+1` then the sampler with which it will be accessed should have MIRROR wrapping on `someComponent`